### PR TITLE
blackbox-terminal: support sixel

### PR DESCRIPTION
**Pull request by linsui**
## Description of changes

Use the same VTE as upstream for sixel support.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by chuangzhu — APPROVED** [review]
Result of `nixpkgs-review pr 248682` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>blackbox-terminal</li>
  </ul>
</details>

File: pkgs/applications/terminal-emulators/blackbox-terminal/default.nix
Comment (kirillrdy) [inline pkgs/applications/terminal-emulators/blackbox-terminal/default.nix:66]: should this be added do vte-gtk4 as an option ?

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: Not sure if it makes sense but added anyway.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: sorry, that's not what I mean.

I was suggesting to add sixel option to vte-gtk4

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: https://github.com/NixOS/nixpkgs/issues/252845

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: thank you for the link, after reading more about

- blackbox still consider sixel experimental
- blackbox doesn't require unstable vte https://gitlab.gnome.org/raggesilver/blackbox/-/blob/main/src/meson.build?ref_type=heads#L45

I feel like blackbox should keep using stable vte.

usually it would be ok for have an optional feature, but the fact that it requires different version of vte sound like bad idea.

maybe someone is willing to maintain patches against stable version of vte ?

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: Blackbox doesn't "require" the unstable vte. So currently blackbox works. But the sixel support requires unstable vte. And the official flatpak does use an unstable vte and I want to use the same vte. https://gitlab.gnome.org/raggesilver/blackbox/-/blob/main/com.raggesilver.BlackBox.json#L36

> but the fact that it requires different version of vte sound like bad idea.

Not sure what "different" means here. The version is different from the stable version but it's the same as upstream.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: different meaning, nixpkgs currently has vte version 0.72.2 and this PR is adding version 3c8f66be867a which is development branch of 0.73.0

```
git diff 3c8f66be867aca6656e4109ce880b6ea7431b895..0.72.2 | wc -l
8690
```


what this means, its fairly easy to make change to vte derivation and break blackboxe-terminal version of vte, eg patches that apply cleanly to vte 0.72.2 but not 0.73.0 branch.

this is essentially equivalent to maintaining two versions of vte in nixpkgs.

my advise is to wait until 0.74 lands https://github.com/NixOS/nixpkgs/pull/247766

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > my advise is to wait until 0.74 lands https://github.com/NixOS/nixpkgs/pull/247766

You haven't get the point. :) We don't need 0.73, we need development branch. Sixel support is not in release versions.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: ok, i see they removed sixel support from stable branch

```
commit e0720853f59ebf54e022b4cd164de2d9b8a21fc1
Author: Christian Persch <chpe@src.gnome.org>
Date:   Sat Aug 5 19:29:08 2023 +0200

    all: Remove SIXEL support from stable branch
```

options in no particular order

- stick with current approach in this PR, but make it optional so that default blackbox-terminal less likely to have broken build
- maintain separate unstable vte derivations that is different from stable vte, therefore also less likely to break
- wait until sixel support is in stable so that we only maintain one version of vte

also, if its unclear to what type of breakages I'm referring to. I picked a first commit after removal of sixel and tried applying it on top of 3c8f66be867aca6656e4109ce880b6ea7431b895 and it failed to apply.  This is the most obvious breakage , but essentially any change made to vte derivation might not work with version of vte in this PR.

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > stick with current approach in this PR, but make it optional so that default blackbox-terminal less likely to have broken build

You want me to flip the default value?

> maintain separate unstable vte derivations that is different from stable vte, therefore also less likely to break

You want me to vendor the whole nix file instead of overriding it?

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: 1) yes, that would mean by default to use stable vte
2) I mean for nixpkgs to maintain two versions of vte. If by  "vendoring"  you mean for backbox-terminal have a full copy of vte derivation, then no, that's not what I meant


one more note about option 1), it would be better instead of sixel option to change version of vte, but instead apply sixel patches to current stable version. 

this makes me reconsider options into

- maintain sixel patches against stable vte
- maintain separate unstable vte ( not in backbox-terminal but in nixpkgs) 
- wait until sixel is in stable

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: I thought it's the worst choice. A override is much more robust that a patch.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: override in this PR introduces an other version of vte, this means any changes, to vte derivation will require testing/building both versions.

that's extra workload to vte maintainers ( gnome team )  or it would just mean that blackbox-terminal will more likely to fail to build each time vte is changed

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > maintain sixel patches against stable vte

This requires too much maintaince and more likely causes breakage.

> maintain separate unstable vte ( not in backbox-terminal but in nixpkgs)

https://github.com/NixOS/nixpkgs/pull/224104 That was rejected.

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: How can I apply the patch without overriding?

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > How can I apply the patch without overriding?

sorry, I am not sure I understand your question. is this in continuation of 

> maintain sixel patches against stable vte

?

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: I don't understand why you thought it's better to use a patch. How can you apply the patch without overriding?

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: it's better to use a patch, so that blackbox-terminal with sixel use roughly same version of as blackbox-terminal without sixel, it would be very confusing if changing  one option bring/changes other behaviors.

patch would live inside current vte derivation. Maintaining that patch, as you have pointed out will require work, potentially with each vte update, but at least regular blackbox-terminal less likely to break builds because it will be using vte that is maintained by nixpkgs

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: sorry to jump into an other reply


>     maintain sixel patches against stable vte
those patches might not get accepted into vte derivation, then they can live in blackbox-terminal derivation

>     maintain separate unstable vte ( not in backbox-terminal but in nixpkgs)
this is most likely not to get accepted

>     wait until sixel is in stable
i think from nixpkgs point of view this is the simpler option.

people can override version of vte on their local machines all they want, but for nixpkgs it would make sense to wait unti upstream accepts those features ( eg not remove them from stable )

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > it would be very confusing if changing one option bring/changes other behaviors.

So I thought we shouldn't provide the option. If we just use the same vte as upstream, there is no confusion.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: after reading https://gitlab.gnome.org/raggesilver/blackbox/-/issues/273 I am not convinced that for nixpkgs we should wait until its in stable vte. Those that want can try it today, even this PR can be useful for people looking for sixel support

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: Yes, we should wait until stable vte supports sixel. We just need to build blackbox against the vte version choosed by the blackbox dev.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: haha I am so sorry for typo :-), please read previous post again

https://gitlab.gnome.org/raggesilver/blackbox/-/issues/273#note_1768201
as per that comment, only Flatpack is built with that version, and if people start reporting sixel issues to vte then sixel will be removed from blackbox

clear indication that it's too early to ship it in nixpkgs

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > only Flatpack is built with that version

IIUC, only Flatpak version is built with sixel support because the blackbox dev can only control that. And other distros can't use a different version of vte only for blackbox. In nixpkgs we don't need to ship it globally. We only ship it to blackbox users behind an experiment option. It's the choice of upstream.

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: ok. with this in mind

how about ?

- make blackbox-terminal always use version of vte that's listed in Flatpack
- make sixel support optional flag, that's disabled by default
- need approval from blackbox-terminal maintainer

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: I flipped the option.

File: pkgs/applications/terminal-emulators/blackbox-terminal/default.nix
Comment (kirillrdy) [inline pkgs/applications/terminal-emulators/blackbox-terminal/default.nix:67]: option should not affect src, but only sixel support

Comment (linsui) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix:67]: You mean we should always use the same src?

Comment (kirillrdy) [reply pkgs/applications/terminal-emulators/blackbox-terminal/default.nix]: > 
>     make blackbox-terminal always use version of vte that's listed in Flatpack
>     make sixel support optional flag, that's disabled by default
>     need approval from blackbox-terminal maintainer
>

**Review by kirillrdy — APPROVED** [review]
@linsui appreciate you patience.

@chuangzhu can you have a look at recent changes ?

**Comment by chuangzhu** [comment]
> need approval from blackbox-terminal maintainer

I don't have a strong opinion against this.

> this comes with higher chance of blackbox-terminal build fails when vte derivation changes

@linsui maybe add yourself as a maintainer and help tracking it?

**Comment by linsui** [comment]
No problem, added.

**Review by FliegendeWurst — COMMENTED** [review]
Your last commit is not [associated with your Github account](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user).

**Review by chuangzhu — APPROVED** [review]

**Comment by wyndon** [comment]
Can we get this merged ?

**Review by NickCao — APPROVED** [review]

**Comment by linuxmobile** [comment]
Waiting this pr long time

**Pull request merged**
